### DISCUSSION TITLE
Mutations without sending data to the server?

### DISCUSSION THREAD

**Discussion by carloslaborde** [reduxjs/redux-toolkit#2790] [Q&A] [answered]
After downloading data (say a JSON object) via RTK Query, if we want users to be able to modify that data locally, and then confirm before updating the server with the mutated data, is the best practice to use RTK Query mutation, or to duplicate it in a separate, local-only (non-RTK-Query) Redux store, and then perform a server mutation via RTK when it's ready?

I'd kind of hate to duplicate it, because it's already in the store (albeit in a less directly-accessible way), but this temporary "WIP" state is necessary (i.e. without being uploaded).

I couldn't find anything obvious after some searching & reading the docs, but I'm guessing someone's encountered a case like this, any thoughts?

**Comment by markerikson** [comment, marked as answer]
I'd definitely keep local WIP data separately if I personally was doing it, yeah.  That's basically what I did wayyyy back in 2016 on my first Redux project, and later wrote about in a tutorial post:

- https://blog.isquaredsoftware.com/2017/01/practical-redux-part-8-form-draft-data-management/

That keeps a clear separation between "this is the data that matches what's on the server", and "here's what the user is currently editing".

**Reply by carloslaborde** [reply]
Amazing. Very helpful breakdown of the core principle backing the decision: **whenever possible, represent your application's behavior through state**.

**Comment by taochu** [comment]
@markerikson 

In this situation, do you recommend using `api.endpoints.xxx.matchFulfilled` to put data from RTK query into the store or is there another way that you think we should do instead?

**Reply by phryneas** [reply]
I'd suggest to keep your slice empty and just save there what changed from the server data.
